### PR TITLE
Refactor modules support

### PR DESCRIPTION
**Pull request by waruqi**
https://github.com/xmake-io/xmake/pull/4321
https://github.com/xmake-io/xmake/issues/4629

### CODE REVIEW

File: xmake/rules/c++/modules/modules_support/dependency_scanner.lua [hunk 1]
```diff
@@ -593,145 +447,104 @@ function fallback_generate_dependencies(target, jsonfile, sourcefile, preprocess
...
+-- removed named module not used in the translation units
+-- when building a library we only cull external modules because we need module objectfiles to be linked inside the library
+-- on an executable we cull explicitly referenced module
+function cull_unused_modules(target, modules)
 
-function get_metafile(target, modulefile)
-    local outputdir = get_outputdir(target, modulefile)
-    return path.join(outputdir, path.filename(modulefile) .. ".meta-info")
-end
+    local cull_all_modules = target:kind() == "executable"
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/dependency_scanner.lua:515]: no `executable` kind, only binary and `target:is_binary()`

File: xmake/rules/c++/modules/modules_support/dependency_scanner.lua [hunk 2]
```diff
@@ -593,145 +447,104 @@ function fallback_generate_dependencies(target, jsonfile, sourcefile, preprocess
...
+        --     if module.requires then
+        --         for required, _ in pairs(module.requires) do
+        --             table.insert(needed_modules, required)
+        --         end
+        --     end
+        --     if table.find(needed_modules, name) then
+        --        culled[objectfile] = module
+        --     end
+        -- else
+             culled[objectfile] = module
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/dependency_scanner.lua:545]: cull_unused_modules just return the original modules. if it does not work, we should remove it.

or 

``lua
function cull_unused_modules(target, modules)
    return modules
```

@Arthapz

**Comment by Arthapz** [comment]
https://github.com/xmake-io/xmake/pull/4321#issuecomment-1922625604
> > i got clang support to use compinst:compile, but now for msvc i need a way to have compinst:compile omit -Fo flag
> 
> Why omit it? Can you show me the whole compile command and flags?

```
cl.exe -nologo -O2 -fp:fast -std:c++20 -DNDEBUG -TP -ifcOutput build\.gens\cpp_with_moduledeps\win
dows\x64\release\rules\bmi\cache\modules\0ca5c33d\mod.ifc -interface -ifcOnly -c src\mod.mpp
```

-ifcOnly disable generation of .o file but after some tests, finally we can add the -Fo flag, it will just be an no-op

**Comment by Arthapz** [comment]
i think we should introduce moduleonly target kind which will behave like headeronly but generate module deps and generate module metadata

**Comment by waruqi** [comment]
Don't think about introducing too much kind for now, it would affect a lot of things. Use existing kind and custom rule implementations wherever possible.

**Comment by Arthapz** [comment]
> Don't think about introducing too much kind for now, it would affect a lot of things. Use existing kind and custom rule implementations wherever possible.

ok i'll try to implement it with a rule for headeronly targets

**Comment by waruqi** [comment]
https://github.com/xmake-io/xmake/pull/4727 

Modules incremental compilation does not work.

**Comment by Arthapz** [comment]
I don't have much Time today, so i'll check tomorrow

**Comment by Arthapz** [comment]
i may have fixed incremental compilation in my module reuse PR, so i'll create the PR, but it is based on https://github.com/xmake-io/xmake/pull/4707

**Comment by waruqi** [comment]
> i may have fixed incremental compilation in my module reuse PR #4728, so i'll create the PR, but it is based on #4707

packages test does not work.

```console
ruki@0d35ffa77272:/mnt/xmake/tests/projects/c++/modules/packages$ xmake
[ 22%]: <packages> compiling.module.release bar
[ 22%]: <packages> compiling.module.release foo
[ 22%]: <packages> compiling.module.release bar2
[ 66%]: compiling.release src/main.cpp
[ 77%]: linking.release packages
[100%]: build ok, spent 1.373s
ruki@0d35ffa77272:/mnt/xmake/tests/projects/c++/modules/packages$ xmake
[ 22%]: <packages> compiling.module.release bar2
[ 22%]: <packages> compiling.module.release foo
[ 22%]: <packages> compiling.module.release bar
[ 66%]: compiling.release src/main.cpp
[ 77%]: linking.release packages
[100%]: build ok, spent 1.381s
```

**Pull request merged**